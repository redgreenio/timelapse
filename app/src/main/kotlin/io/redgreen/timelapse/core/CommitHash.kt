package io.redgreen.timelapse.core

inline class CommitHash(val value: String)
