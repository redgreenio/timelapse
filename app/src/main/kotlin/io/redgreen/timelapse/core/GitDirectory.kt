package io.redgreen.timelapse.core

import java.io.File
import java.util.Optional
import org.eclipse.jgit.lib.Constants.HEAD
import org.eclipse.jgit.lib.RepositoryBuilder

/* Not using a sealed class here because it can still expose a private constructor via copy constructors. Therefore,
 * we are also responsible for implementing [#hashCode], [#equals], and [#toString] functions. */
class GitDirectory private constructor(val path: String) {
  companion object {
    private const val GIT_DIRECTORY_NAME = ".git"

    fun from(path: String): Optional<GitDirectory> {
      return if (isGitDirectoryCheck(path)) {
        Optional.of(GitDirectory(path))
      } else {
        Optional.empty()
      }
    }

    private fun isGitDirectoryCheck(path: String): Boolean {
      val head = RepositoryBuilder().setGitDir(File(path)).build().resolve(HEAD)
      return head != null
    }
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (javaClass != other?.javaClass) return false

    other as GitDirectory

    if (path != other.path) return false

    return true
  }

  override fun hashCode(): Int {
    return path.hashCode()
  }

  override fun toString(): String {
    return "GitDirectory(path='$path')"
  }
}
