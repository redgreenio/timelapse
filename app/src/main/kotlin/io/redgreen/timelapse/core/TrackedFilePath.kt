package io.redgreen.timelapse.core

/** Represents a file path tracked by Git. */
inline class TrackedFilePath(val value: String)
