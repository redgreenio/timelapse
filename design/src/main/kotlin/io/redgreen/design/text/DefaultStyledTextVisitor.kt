package io.redgreen.design.text

open class DefaultStyledTextVisitor : StyledTextVisitor {
  override fun onEnterLine(lineNumber: Int) {
    /* override if necessary :) */
  }

  override fun onExitLine(lineNumber: Int) {
    /* override if necessary :) */
  }

  override fun onEnterLine(lineNumber: Int, style: LineStyle) {
    /* override if necessary :) */
  }
}
