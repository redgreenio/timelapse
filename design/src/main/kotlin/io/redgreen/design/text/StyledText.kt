package io.redgreen.design.text

import java.lang.StringBuilder
import java.util.Optional

data class StyledText(val text: String) {
  private val lineStyles = mutableListOf<LineStyle>()

  fun visit(visitor: StyledTextVisitor) {
    var lineNumber = 1

    getLineStyle(lineNumber)
      .ifPresentOrElse(
        { visitor.onEnterLine(lineNumber, it) },
        { visitor.onEnterLine(lineNumber) }
      )

    val textBuilder = StringBuilder()

    text.onEachIndexed { index, char ->
      if (char == '\n') {
        lineNumber++

        getLineStyle(lineNumber)
          .ifPresentOrElse(
            { visitor.onEnterLine(lineNumber, it) },
            { visitor.onEnterLine(lineNumber) }
          )
      } else {
        if (char != '\n') {
          textBuilder.append(char)
        }
      }

      if ((index + 1 > text.lastIndex) || (index + 1 != text.lastIndex && text[index + 1] == '\n')) {
        visitor.onText(textBuilder.toString())
        textBuilder.clear()

        getLineStyle(lineNumber)
          .ifPresentOrElse(
            { visitor.onExitLine(lineNumber, it) },
            { visitor.onExitLine(lineNumber) }
          )
      }
    }
  }

  private fun getLineStyle(lineNumber: Int): Optional<LineStyle> {
    return Optional.ofNullable(lineStyles.find { lineNumber in it.lineNumberRange })
  }

  fun addStyle(lineStyle: LineStyle): StyledText {
    lineStyles.add(lineStyle)
    return this
  }
}
