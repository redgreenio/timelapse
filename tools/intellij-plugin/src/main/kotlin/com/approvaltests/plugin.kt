package com.approvaltests

const val ID = "io.redgreen.timelapse.dev"
const val DISPLAY_NAME = "Approval Tests"
