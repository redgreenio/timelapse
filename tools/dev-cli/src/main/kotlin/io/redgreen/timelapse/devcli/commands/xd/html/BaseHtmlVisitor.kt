package io.redgreen.timelapse.devcli.commands.xd.html

import io.redgreen.design.text.LineStyle
import io.redgreen.design.text.StyledTextVisitor
import io.redgreen.design.text.TextStyle
import kotlin.LazyThreadSafetyMode.NONE

class BaseHtmlVisitor : StyledTextVisitor {
  companion object {
    private const val INDENT = "  "
    private const val NEWLINE = "\n"

    private const val MARKER_TABLE_ROWS = "{table-rows}"
  }

  private val contentBuilder = StringBuilder()

  private val template by lazy(NONE) {
    BaseHtmlVisitor::class.java.classLoader
      .getResourceAsStream("template.html")!!
      .reader()
      .readText()
  }

  val content: String
    get() = template.replace(MARKER_TABLE_ROWS, contentBuilder.toString())

  override fun onText(text: String) {
    contentBuilder.append(text)
  }

  override fun onEnterLine(lineNumber: Int) {
    if (lineNumber != 1) {
      contentBuilder.append(NEWLINE)
    }
    contentBuilder
      .append("<tr>")
      .append(NEWLINE)
      .append("""$INDENT<td class="line-number">$lineNumber</td>""")
      .append(NEWLINE)
      .append("$INDENT<td>")
  }

  override fun onExitLine(lineNumber: Int) {
    contentBuilder
      .append("</td>")
      .append(NEWLINE)
      .append("</tr>")
  }

  override fun onEnterLine(lineNumber: Int, lineStyle: LineStyle) {
    // no-p
  }

  override fun onExitLine(lineNumber: Int, lineStyle: LineStyle) {
    // no-p
  }

  override fun onBeginStyle(textStyle: TextStyle) {
    // no-p
  }

  override fun onEndStyle(textStyle: TextStyle) {
    // no-p
  }
}
