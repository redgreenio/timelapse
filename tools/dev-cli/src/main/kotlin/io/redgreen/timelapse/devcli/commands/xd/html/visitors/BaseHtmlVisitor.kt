package io.redgreen.timelapse.devcli.commands.xd.html.visitors

import io.redgreen.design.text.LineStyle
import io.redgreen.design.text.StyledTextVisitor
import io.redgreen.design.text.TextStyle
import io.redgreen.timelapse.devcli.commands.TOOL_VERSION
import org.apache.commons.text.StringEscapeUtils
import kotlin.LazyThreadSafetyMode.NONE

class BaseHtmlVisitor(
  private val title: String,
  private val affectedLineNumbers: List<Int> = emptyList()
) : StyledTextVisitor {
  companion object {
    private const val INDENT = "  "
    private const val NEWLINE = "\n"
    private const val CHAR_SPACE = ' '
    private const val CHAR_TAB = '\t'

    private const val NBSP = "&nbsp;"
    private const val TAB = "$NBSP$NBSP$NBSP$NBSP"

    private const val MARKER_TITLE = "{title}"
    private const val MARKER_TABLE_ROWS = "{table-rows}"
    private const val MARKER_VERSION = "{tool-version}"
  }

  private val contentBuilder = StringBuilder()

  private val template by lazy(NONE) {
    BaseHtmlVisitor::class.java.classLoader
      .getResourceAsStream("template.html")!!
      .reader()
      .readText()
  }

  val content: String
    get() {
      val rawTableRows = contentBuilder.toString()
      val indentedTableRows = rawTableRows.split(NEWLINE).joinToString(NEWLINE) { "$INDENT$it" }

      return template
        .replace(MARKER_TITLE, title)
        .replace(MARKER_TABLE_ROWS, indentedTableRows)
        .replace(MARKER_VERSION, TOOL_VERSION)
    }

  override fun onText(text: String) {
    contentBuilder.append(toHtmlFriendly(text))
  }

  override fun onEnterLine(lineNumber: Int) {
    if (lineNumber != 1) {
      contentBuilder.append(NEWLINE)
    }
    contentBuilder
      .append("<tr>")
      .append(NEWLINE)
      .append("""$INDENT<td class="line-number">$lineNumber</td>""")
      .append(NEWLINE)

    val muted = lineNumber !in affectedLineNumbers
    if (muted) {
      contentBuilder.append("""$INDENT<td class="muted">""")
    } else {
      contentBuilder.append("""$INDENT<td>""")
      if (isInsideMultilineString) {
        contentBuilder.append("<span class=\"string\">")
      }
    }
  }

  override fun onExitLine(lineNumber: Int) {
    if (isInsideMultilineString) {
      contentBuilder.append("</span>")
    }

    contentBuilder
      .append("</td>")
      .append(NEWLINE)
      .append("</tr>")
  }

  override fun onEnterLine(lineNumber: Int, lineStyle: LineStyle) {
    if (isInsideMultilineString) {
      contentBuilder.append("<span class=\"string\">")
    }
  }

  override fun onExitLine(lineNumber: Int, lineStyle: LineStyle) {
    if (isInsideMultilineString) {
      contentBuilder.append("</span>")
    }
  }

  private var isInsideMultilineString = false

  override fun onBeginStyle(textStyle: TextStyle) {
    if (textStyle.name == "open-multiline-string") {
      isInsideMultilineString = true
    }

    if (textStyle.name == "close-multiline-string") {
      isInsideMultilineString = false
    }

    if (textStyle.name != "end-string") {
      contentBuilder.append("""<span class="${textStyle.name}">""")
    }
  }

  override fun onEndStyle(textStyle: TextStyle) {
    if (textStyle.name != "string") {
      contentBuilder.append("</span>")
    }

    if (textStyle.name == "open-multiline-string" && isInsideMultilineString) {
      contentBuilder.append("<span class=\"string\">")
    } else if (textStyle.name == "close-multiline-string") {
      contentBuilder.append("</span>")
    }
  }

  private fun toHtmlFriendly(line: String): String {
    val startSpaceCharsTrimmedLine = line.trimStart(CHAR_SPACE)
    val nStartSpaceChars = line.length - startSpaceCharsTrimmedLine.length

    val htmlEscapedLine = StringEscapeUtils.escapeHtml4(line).trimStart(CHAR_SPACE)
    return (1..nStartSpaceChars)
      .joinToString("") { NBSP } + htmlEscapedLine
      .replace("$CHAR_TAB", TAB)
  }
}
