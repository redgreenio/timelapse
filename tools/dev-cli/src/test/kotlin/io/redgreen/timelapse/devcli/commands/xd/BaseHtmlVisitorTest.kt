package io.redgreen.timelapse.devcli.commands.xd

import io.redgreen.design.text.StyledText
import io.redgreen.timelapse.devcli.commands.xd.html.visitors.BaseHtmlVisitor
import org.approvaltests.Approvals
import org.approvaltests.core.Options
import org.approvaltests.scrubbers.RegExScrubber
import org.junit.jupiter.api.Test

class BaseHtmlVisitorTest {
  private val toolVersionScrubber = RegExScrubber(
    "<!--Generated by dev-cli: v\\d+.\\d+.\\d+(-\\w+)?-->",
    "<!--Generated by dev-cli: v[tool-version]-->"
  )
  private val scrubberOptions = Options(toolVersionScrubber)

  @Test
  fun `it should generate a base HTML for Xd experimentation`() {
    // given
    val kotlinSource = """
      fun add(x: Int, y: Int) {
        return x + y
      }
      
    """.trimIndent()
    val styledText = StyledText(kotlinSource)
    val visitor = BaseHtmlVisitor("Untitled")

    // when
    styledText.visit(visitor)

    // then
    Approvals.verifyHtml(visitor.content, scrubberOptions)
  }

  @Test
  fun `it should highlight affected lines in a commit`() {
    // given
    val kotlinSource = """
      fun add(x: Int, y: Int) {
        return x + y
      }
      
    """.trimIndent()
    val styledText = StyledText(kotlinSource)
    val visitor = BaseHtmlVisitor("Untitled", listOf(2))

    // when
    styledText.visit(visitor)

    // then
    Approvals.verifyHtml(visitor.content, scrubberOptions)
  }
}
