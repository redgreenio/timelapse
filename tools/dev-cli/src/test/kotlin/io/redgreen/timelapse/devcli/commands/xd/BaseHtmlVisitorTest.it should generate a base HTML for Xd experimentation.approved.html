<!DOCTYPE html>
<html lang="en-US">

<head>
  <style>
    @import url('https://fonts.googleapis.com/css2?family=JetBrains+Mono&display=swap');
  </style>
  <style>
    body {
      font-family: 'JetBrains Mono', monospace;
      font-size: small;
      margin-left: 0;
      margin-right: 0;
    }

    table {
      border-collapse: collapse;
    }

    tr td {
      padding-left: 0.5em;
    }

    .line-number {
      color: #9F9F9F;
      min-width: 3em;
    }
  </style>
  <title>{title}</title>
</head>

<body>
<table>
  <tbody>
  <tr>
    <td class="line-number">1</td>
    <td>fun add(x: Int, y: Int) {</td>
  </tr>
  <tr>
    <td class="line-number">2</td>
    <td>  return x + y</td>
  </tr>
  <tr>
    <td class="line-number">3</td>
    <td>}</td>
  </tr>
  </tbody>
</table>
</body>
</html>
